File: c++/reverseBetween.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode* buildList(int a[], int n);

ListNode *reverseBetween(ListNode *head, int m, int n) {
    if (!head || n-m == 0) return head;
    ListNode *prev, *curr, *futu, *start, *preend;
    int cntn = n-1;
    if (m == 1) {    
        start = head;
        prev = head;
        curr = head->next;
    } else  {    // m > 1
        int cntm = m-1;
        prev = head; curr = head->next; --cntm;
        while (cntm) {
            prev = curr;
            curr = curr->next;
            --cntn;
            --cntm;
        }
        preend = prev;
        start = curr;
    }
    while(cntn) {
        futu = curr->next;
        curr->next = prev;
        prev = curr;
        curr = futu;
        --cntn;
    }

    if (m == 1) {
        head = prev;
        start->next = curr;
    } else  {
        preend->next = prev;
        start->next = curr;
    }
    return head;
}

int main() {
    int n = 5;
    int a[] = {1, 2, 3, 4, 5};
    ListNode* head = buildList(a, n);
    cout << endl;
    ListNode* curr = head;
    while(curr)  {
        cout << curr->val << ", ";
        curr = curr->next;
    }

    ListNode* result = reverseBetween(head, 2, 5);

    cout << endl;
    while(result)  {
        cout << result->val << ", ";
        result = result->next;
    }
    return 0;
}

ListNode* buildList(int a[], int n) {
    ListNode* head, *curr;
    for (int i = 0; i < n; ++i) {
        if (i == 0)  {
            head = new ListNode(a[i]);
            curr = head;
        } else  {
            curr->next = new ListNode(a[i]);
            curr = curr->next;
        }
    }
    return head;
}

File: java/reverseBetween.java
import com.ListNode;

import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

public class reverseBetween {
    public static class Solution {
        public ListNode reverseBetween(ListNode head, int m, int n) {
            if (head == null || head.next == null || m == n) return head;
            ListNode curr = head;
            ListNode prev = null;
            ListNode tail = null;
            int cnt = 1;
            while (curr != null && cnt < m) {
                prev = curr;
                curr = curr.next;
                ++cnt;
            }

            ListNode btai = curr;
            if (m == 1) {
                tail = null;
            } else {
                tail = prev;
                tail.next = null;
            }
            ListNode bcur = curr;
            curr = curr.next; // ++cnt;
            bcur.next = null;
            ListNode fast = null;
            ++cnt;
            while (curr != null && cnt < n) {
                fast = curr.next;
                curr.next = bcur;
                bcur = curr;
                curr = fast;
                ++cnt;
            }
            fast = curr.next;
            curr.next = bcur;
            btai.next = fast;
            if (tail != null) {
                tail.next = curr;
                return head;                
            } else
                return curr;
        }
    }

    public static void main(String[] args){
        Solution result = new Solution();
        int [] a = {3, 5};  // special situation
        ListNode head = new ListNode(a[0]);
        head.buildList(head, a);
        head.printList(head);
        ListNode res = result.reverseBetween(head, 1, 2);

        System.out.println("");
        while (res != null) {
            System.out.print(res.val + ", ");
            res = res.next;
        }
    }
}
